### ISSUE TITLE
Allow groups to match items non-greedy?

### ISSUE THREAD

**Issue by simendsjo** [alphapapa/org-super-agenda#153]
Before I go further with this and create a PR, I want to get your input on whether you want it and your thoughts on the design.

This is related to my previous issue, https://github.com/alphapapa/org-super-agenda/issues/152.
On further investigation, I found the problem was that each item could only exist in a single group. This means I cannot both have a "schedule" group and a list of upcoming deadlines as an item can be both, and would get caught by the first. The same issue applies in many cases. I started creating several different functions for different `org-super-agenda-groups` permutations, but it became difficult to create groups which didn't overlap, and just creating all of these felt wrong.

Looking at the code made it seem like a small change to allow an item to both match, and still flow to groups further down. A non-greedy match.

I created a simple implementation for this so I could make sure items matches several groups, and for my use-case, this works fine.

``` patch
diff --git a/org-super-agenda.el b/org-super-agenda.el
index 7878c40..541b0dd 100644
--- a/org-super-agenda.el
+++ b/org-super-agenda.el
@@ -121,7 +121,7 @@
 ;;;; Variables
 
 (defconst org-super-agenda-special-selectors
-  '(:name :order :face :transformer)
+  '(:name :order :face :transformer :forward)
   ;; This needs to be manually updated if any are added.
   "Special, non-grouping selectors.")
 
@@ -1024,7 +1024,7 @@ see."
            ;; This is the implicit OR
            append matching into all-matches
            and collect auto-section-name into names
-           and do (setq items non-matching)
+           and do (setq items (append non-matching (and (memq :forward group) matching)))
            for name = (if (stringp (car names))
                           (s-join " and " (-non-nil names))
                         ;; Probably an :auto-group
```

What do you think about a feature like this? I would like to avoid maintaining a fork, so I'll happily do some work on this. Note that I don't know much lisp or emacs, so it might take a while.

**Comment by alphapapa** [comment]
This is an interesting idea.  I think at least one other user has expressed interest in it, so it could be useful.  I'm willing to consider it.

Regarding `"Other items"`, it would probably be better to just use the `:discard` selector as the last group.

**Comment by simendsjo** [comment]
> This is an interesting idea. I think at least one other user has expressed interest in it, so it could be useful. I'm willing to consider it.

Ok, I'll make a proposal.

> Regarding "Other items", it would probably be better to just use the :discard selector as the last group.

This doesn't work as it will collect everything that has been forwarded. I'm not able to really discard them it seems.
``` lisp
          (:name "Blocked"
           :todo "FEEDBACK")
          (:name "empty?" :anything :discard)
```
Here I get a very big "empty?" group that includes everything that has flowed through the "Blocked", which is everything except "FEEDBACK" when i use :forward. So this doesn't discard them without collecting them to the group it seems.

**Comment by alphapapa** [comment]
```
(:discard (:anything t))
```

**Comment by cgoetzke** [comment]
For what it's worth, I'd use this if it gets submitted & merged.

**Comment by barryridge** [comment]
+1 for this.  It seems to me that this is the way it should be by default and it was actually very confusing, to me anyway, that items wouldn't show up in multiple groups grouped by tags and it took me a while to figure out what was going on.  That seems antithetical to the way in which tags, at least, are normally used.

**Comment by alphapapa** [comment]
I think the implementation of this would be fairly simple, as @simendsjo showed.  The hard part may be choosing the best name for the new selector.  Some possibilities:

* `:forward`
* `:copy`
* `:share`
* `:duplicate`
* `:borrow`
* `:no-consume`
* `:non-consuming`

Suggestions welcome!

**Comment by simendsjo** [comment]
> I think the implementation of this would be fairly simple, as @simendsjo showed. The hard part may be choosing the best name for the new selector. Some possibilities:

Yeah, that's pretty much impossible :) I like no-consume/non-consuming (or maybe non-greedy?)

**Comment by alphapapa** [comment]
I think I like `:copy` the best: it's short, easy to type, and I think it's understood easily enough (i.e. it copies items into the group, implying that the original exists elsewhere).

I think `:non-greedy` could be confusing because of its role in regexp jargon.

WDYT?  Thanks.

**Comment by simendsjo** [comment]
> I think I like `:copy` the best: it's short, easy to type, and I think it's understood easily enough (i.e. it copies items into the group, implying that the original exists elsewhere).

Still like `non-consuming` :) The fact that the default is greedy/consuming was unintuitive for me, so a `non-consuming` will both say that the rule shouldn't consume the item, as well as it consuming in the default case. `copy` doesn't seem as self-describing. But I guess five different people will have five different opinions on this.

**Comment by barryridge** [comment]
Just to throw the cat amongst the pigeons here - my inkling is that most users trying `org-super-agenda` out for the first time would be expecting the "non-greedy" behavior to be the default behavior and would be confused to find that this is not the case (as a couple of us here already have been).  Would it be possible to make "non-greedy" the default behavior and then use a selector to change it to "greedy"?  And my suggestion for the selector name would then be `:greedy` - that would be the most user-friendly implementation IMHO :)

**Comment by alphapapa** [comment]
I think that expectations vary depending on the user's prior understanding of Org Agenda.  IME, hearing from users of this package, users' knowledge level (regarding Org, Org Agenda, Elisp, etc) varies widely.  I don't know what is the best default for all users.

AFAIK, generally Org Agenda shows each item once.  So if non-consuming were the default behavior, that would mean duplicating items, contrary to default Agenda behavior.  That could confuse users, because it could make it appear that one item were multiple items.  (Imagine if there were a dired-super-groups package for Dired, and non-consuming were the default behavior for its grouping: it could make it appear that multiple copies of a file existed in a directory.)

And at this point, changing the default would be a significant change for existing users of this package; I expect that I'd have to answer many of their questions about why their agendas were suddenly larger and had duplicate items.  

So I don't think changing the default would be a good idea.  However, we could probably add an option to change the default behavior to whichever the user prefers.

I agree that `:copy` is not a great name for the property.  `:copy-into` might be more descriptive, although also more verbose.  I'd really like to find a concise, one-word description.

**Comment by alphapapa** [comment]
What about just `:consume`, and to disable "greedy"/consuming behavior, the user would use `:consume nil`?

**Comment by barryridge** [comment]
@alphapapa I think you're right to be concerned about changing the default behavior now after the fact - probably best not to upset the apple cart.

I think what you've converged to with `:consume nil` sounds good!  Would be great if you could also set it globally or individually per-group.

**Comment by alphapapa** [comment]
> Would be great if you could also set it globally or individually per-group.

Yes, that's the idea.

Okay, so `:consume` it is.

**Comment by pkazmier** [comment]
I know I’m late to the party but I like `:consume` too.

**Comment by simendsjo** [comment]
Seems like consensus has been reached. I tried taking a stab at this, but as it's quite different from my original proposal, I'm not sure how to add it myself - I'm terrible at elisp.

**Comment by wuqui** [comment]
Are there any news on this? I would love to have this feature, too.

**Comment by simendsjo** [comment]
> Are there any news on this?
> 
> I would love to have this feature, too. It would open a whole range of new applications; for example, I would like to try Gantt-like overviews of coming weeks and months.

I'm still using my old changes from my own branch: https://github.com/alphapapa/org-super-agenda/compare/master...simendsjo:org-super-agenda:simendsjo-experiments
